libs/node-custom/src/lock.ts

```diff
@@ -0,0 +1,46 @@
+/**
+ * A simple lock implementation.
+ */
+export class Lock {
+  private locked = false;
+  private readonly resolvers: Array<() => void> = [];
```

Comment:
I created something similar to this for `print-scan` so I think it would be useful to have in `utils` or `basics`.